### DISCUSSION TITLE
How to use zip.js with create-react-app?

### DISCUSSION THREAD

**Discussion by JosePereira97** [gildas-lormeau/zip.js#258] [Q&A] [answered]
When i load the package using (import * as ZIP from "@zip.js/zip.js") is gives me this error 

./node_modules/@zip.js/zip.js/lib/core/codecs/codec-pool-worker.js 85:72
Module parse failed: Unexpected token (85:72)
File was processed with these loaders:
 * ./node_modules/babel-loader/lib/index.js
You may need an additional loader to handle the result of these loaders.
|
|   if (!workerData.interface) {
>     workerData.worker = new Worker(new URL(workerData.scripts[0], import.meta.url));
|     workerData.worker.addEventListener(MESSAGE_EVENT_TYPE, onMessage, false);
|     workerData.interface = {

i dont know why it happens i tried to read some Q&A in the repository, but i didnt find nothing. Do you guys know the origin of this error? And how to resolve it?

**Comment by gildas-lormeau** [comment]


**Reply by JosePereira97** [reply]
Thank you for you fast response. I am not using any webpack, i just start my app using create-react-app, and npm start? It is necessary to use a webpack for the package to work properly? Ty again for all the help.

**Reply by iquasere** [reply]
Running ```npm install webpack``` did not fix this. Neither for ```webpack@4``` nor for ```webpack@5```. Running ```npm start``` again showed this error message:
```
> mosguito@0.1.0 start C:\Users\User\Desktop\MOSGUITO
> react-scripts start


There might be a problem with the project dependency tree.
It is likely not a bug in Create React App, but something you need to fix locally.

The react-scripts package provided by Create React App requires a dependency:

  "webpack": "4.44.2"

Don't try to install it manually: your package manager does it automatically.
However, a different version of webpack was detected higher up in the tree:

  C:\Users\User\Desktop\MOSGUITO\node_modules\webpack (version: 5.40.0)

Manually installing incompatible versions is known to cause hard-to-debug issues.

If you would prefer to ignore this check, add SKIP_PREFLIGHT_CHECK=true to an .env file in your project.
That will permanently disable this message but you might encounter other issues.

To fix the dependency tree, try following the steps below in the exact order:

  1. Delete package-lock.json (not package.json!) and/or yarn.lock in your project folder.
  2. Delete node_modules in your project folder.
  3. Remove "webpack" from dependencies and/or devDependencies in the package.json file in your project folder.
  4. Run npm install or yarn, depending on the package manager you use.

In most cases, this should be enough to fix the problem.
If this has not helped, there are a few other things you can try:

  5. If you used npm, install yarn (http://yarnpkg.com/) and repeat the above steps with it instead.
     This may help because npm has known issues with package hoisting which may get resolved in future versions.

  6. Check if C:\Users\Utilizador\Desktop\MOSGUITO\node_modules\webpack is outside your project directory.
     For example, you might have accidentally installed something in your home folder.

  7. Try running npm ls webpack in your project folder.
     This will tell you which other package (apart from the expected react-scripts) installed webpack.

If nothing else helps, add SKIP_PREFLIGHT_CHECK=true to an .env file in your project.
That would permanently disable this preflight check in case you want to proceed anyway.

P.S. We know this message is long but please read the steps above :-) We hope you find them helpful!

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! mosguito@0.1.0 start: `react-scripts start`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the mosguito@0.1.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\User\AppData\Roaming\npm-cache\_logs\2021-06-26T14_00_22_914Z-debug.log
```

**Comment by gildas-lormeau** [comment, marked as answer]
`create-react-app` [uses webpack 4](https://github.com/facebook/create-react-app#philosophy) under the hood to build the application and webpack 4 does not fully support web workers without an additional [loader](https://v4.webpack.js.org/concepts/#loaders).

To fix this issue, you must include the loader [@open-wc/webpack-import-meta-loader](https://github.com/open-wc/open-wc) in your project. You can do this by using for example [react-app-rewired](https://github.com/timarney/react-app-rewired#readme) and following these steps:
1. run `npm install react-app-rewired --save-dev` to install `react-app-rewired`
2. run `npm install @open-wc/webpack-import-meta-loader --save-dev` to install `@open-wc/webpack-import-meta-loader`
3. create a file named `config-overrides.js` in the root folder of your project and paste the following content:
```js
module.exports = function override(config, env) {
    config.module.rules.push({
        test: /@zip\.js\/zip\.js\/lib\/core\/codecs\/codec-pool-worker\.js$/,
        use: { 
            loader: require.resolve('@open-wc/webpack-import-meta-loader') 
        }
    });
    return config;
}
```
4. edit your `package.json` file and replace:
```
"start": "react-scripts start",
"build": "react-scripts build",
"test": "react-scripts test",
```
with:
```
"start": "react-app-rewired start",
"build": "react-app-rewired build",
"test": "react-app-rewired test",
```

**Reply by gildas-lormeau** [reply]
It looks like create-react-app will use webpack 5 soon though, see https://github.com/facebook/create-react-app/issues/9994. That should also solve this issue.

**Reply by iquasere** [reply]
Thank you for the help! Unfortunately, this solution returns a next problem: ```Uncaught ReferenceError: ne is not defined```. Even worse, it cannot be reproduced with ```npm start```, as it works fine in ```localhost```. It is when hosted with GitHub Pages that it doesn't work.

You can reproduce the error in the [current state](https://iquasere.github.io/MOSGUITO) of the application, going to "results", and uploading [this archive](https://github.com/gildas-lormeau/zip.js/files/6724948/MOSCA_results.zip). The result should be the display of the contents of the archive, but nothing happens, and that error is printed in the console.

**Reply by iquasere** [reply]
To see how it should work, you can run
```
git clone https://github.com/iquasere/MOSGUITO.git
cd MOSGUITO
npm install
npm start
```
and doing the same on that local version. It works locally, but not with GitHub Pages

**Reply by gildas-lormeau** [reply]
Indeed, I confirm I can reproduce the issue when the app is built. I don't know precisely its cause. It looks like some code (e.g. helper code used to emulate classes) which should be added by webpack is not present in the web worker code.

You can circumvent the issue by disabling web workers, e.g.
```js
import * as zip from "@zip.js/zip.js";

zip.configure({ useWebWorkers: false });
```

**Reply by iquasere** [reply]
That fixed it, thank you very much!

**Reply by gildas-lormeau** [reply]
The version `2.3.12` fixes the build issue. Now, you can use web workers in applications built with `create-react-app`.

**Reply by huangminsky** [reply]
If a vue app ,not a react app 
Is the same NPM package applicable ？

**Comment by MythicManiac** [comment]
For the record, I am having similar issue when transpiling with babel. Specifically what I'm seeing is
```
SyntaxError: Unexpected token (134:72) while parsing projectdir\node_modules\@zip.js\zip.js\lib\core\codecs\codec-pool-worker.js while parsing file: projectdir\node_modules\@zip.js\zip.js\lib\core\codecs\codec-pool-worker.js
```

My babel presets are `["@babel/preset-env", "@babel/preset-react"]`

**Reply by gildas-lormeau** [reply]
I fixed this issue in the version v2.3.12, see https://github.com/gildas-lormeau/zip.js/releases/tag/v2.3.12

**Comment by puxiao** [comment]
You can also use `@craco/craco` .

1. run `yarn add @craco/craco --dev`

2. run `yarn add @open-wc/webpack-import-meta-loader --dev`

3. create a file named `craco.config.js` in the root folder of your project and paste the following content:

   ```
   module.exports = {
       webpack: {
           configure: (config) => {
               config.module.rules.push({
                   test: /\.js$/,
                   use: { loader: require.resolve('@open-wc/webpack-import-meta-loader') }
               });
               return config
           }
   };
   ```

4. edit you `package.json` file and replace:

   ```diff
   - "start": "react-scripts start",
   - "build": "react-scripts build",
   - "test": "react-scripts test",
   
   + "start": "craco start",
   + "build": "craco build",
   + "test": "craco test",
   ```

**Reply by abduraufsherkulov** [reply]
> You can also use `@craco/craco` .
> 
> 1. run `yarn add @craco/craco --dev`
> 2. run `yarn add @open-wc/webpack-import-meta-loader --dev`
> 3. create a file named `craco.config.js` in the root folder of your project and paste the following content:
>    ```
>    module.exports = {
>        webpack: {
>            configure: (config) => {
>                config.module.rules.push({
>                    test: /\.js$/,
>                    use: { loader: require.resolve('@open-wc/webpack-import-meta-loader') }
>                });
>                return config
>            }
>    };
>    ```
> 4. edit you `package.json` file and replace:
>    ```diff
>    - "start": "react-scripts start",
>    - "build": "react-scripts build",
>    - "test": "react-scripts test",
>    
>    + "start": "craco start",
>    + "build": "craco build",
>    + "test": "craco test",
>    ```

Did it work tho? I am using craco and craco-less is not working after using webpack options.

**Reply by puxiao** [reply]
Yes, it can work.
I have an example: https://github.com/puxiao/cesium-issues-9790
